File: android/app/src/main/java/com/example/hundlename/AIFilterService.java
package com.example.hundlename;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.Log;

public class AIFilterService {
    private static final String TAG = "AIFilterService";
    
    // Filter parameters matching Python code
    private int threshold1 = 100;  // Canny threshold1
    private int threshold2 = 200;  // Canny threshold2  
    private boolean enableColorful = true;
    private boolean isEnabled = false;
    private boolean isInitialized = false;

    public boolean initialize() {
        try {
            // For now, initialize without OpenCV to avoid dependency issues
            isInitialized = true;
            Log.d(TAG, "AI Filter Service initialized (stub mode)");
            return true;
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize AI Filter Service", e);
            return false;
        }
    }

    public void setEnabled(boolean enabled) {
        this.isEnabled = enabled;
        Log.d(TAG, "AI Filter " + (enabled ? "enabled" : "disabled"));
    }

    public void setFilterParams(int threshold1, int threshold2, boolean colorful) {
        this.threshold1 = threshold1;
        this.threshold2 = threshold2;
        this.enableColorful = colorful;
        Log.d(TAG, String.format("Filter params: threshold1=%d, threshold2=%d, colorful=%s", 
                                threshold1, threshold2, colorful));
    }

    /**
     * Apply AI filter to hide face with colorful edge detection
     * Matching Python algorithm: Canny edge detection + colorful line rendering
     */
    public Bitmap applyAIFilter(Bitmap inputBitmap) {
        if (!isEnabled || !isInitialized || inputBitmap == null) {
            return inputBitmap;
        }

        try {
            int width = inputBitmap.getWidth();
            int height = inputBitmap.getHeight();
            
            // Create black background bitmap
            Bitmap resultBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
            Canvas canvas = new Canvas(resultBitmap);
            canvas.drawColor(Color.BLACK);
            
            if (enableColorful) {
                // Simulate colorful edge detection (simplified version)
                Paint paint = new Paint();
                paint.setStrokeWidth(3.0f);
                paint.setStyle(Paint.Style.STROKE);
                
                // Draw colorful edges based on position (simplified algorithm)
                for (int y = 0; y < height; y += 10) {
                    for (int x = 0; x < width; x += 10) {
                        // Get pixel from original image
                        if (x < width && y < height) {
                            int pixel = inputBitmap.getPixel(x, y);
                            int brightness = (Color.red(pixel) + Color.green(pixel) + Color.blue(pixel)) / 3;
                            
                            // Only draw edges for bright areas (simplified edge detection)
                            if (brightness > 100) {
                                // Python algorithm: b = round(math.degrees(math.atan(a))) + 90
                                float angle = (x + y) % 180; // Simplified angle calculation
                                int[] hsvColor = hsvToRgb((int)angle, 255, 255);
                                
                                paint.setColor(Color.rgb(hsvColor[0], hsvColor[1], hsvColor[2]));
                                canvas.drawLine(x, y, x + 5, y + 5, paint);
                            }
                        }
                    }
                }
            }
            
            Log.d(TAG, "AI Filter applied successfully");
            return resultBitmap;
            
        } catch (Exception e) {
            Log.e(TAG, "Error applying AI filter", e);
            return inputBitmap;
        }
    }

    /**
     * HSV to RGB conversion matching Python cv2.cvtColor(HSV2BGR)
     */
    private int[] hsvToRgb(int h, int s, int v) {
        float hNorm = (h % 360) / 360.0f;
        float sNorm = s / 255.0f;
        float vNorm = v / 255.0f;
        
        float c = vNorm * sNorm;
        float x = c * (1 - Math.abs((hNorm * 6) % 2 - 1));
        float m = vNorm - c;
        
        float r = 0, g = 0, b = 0;
        
        int sector = (int)(hNorm * 6);
        switch (sector) {
            case 0: r = c; g = x; b = 0; break;
            case 1: r = x; g = c; b = 0; break;
            case 2: r = 0; g = c; b = x; break;
            case 3: r = 0; g = x; b = c; break;
            case 4: r = x; g = 0; b = c; break;
            default: r = c; g = 0; b = x; break;
        }
        
        return new int[]{
            (int)((r + m) * 255),
            (int)((g + m) * 255),
            (int)((b + m) * 255)
        };
    }

    public void release() {
        isEnabled = false;
        isInitialized = false;
        Log.d(TAG, "AI Filter Service released");
    }

    public boolean isEnabled() {
        return isEnabled;
    }
}

File: ios/Runner/Services/AIFilterService.swift
import Foundation
import CoreImage
import UIKit
import Accelerate

class AIFilterService {
    
    // Singleton instance
    static let shared = AIFilterService()
    
    // Filter parameters
    var threshold1: CGFloat = 100.0
    var threshold2: CGFloat = 200.0
    var enableColorful: Bool = true
    
    // Core Image filters
    private let context = CIContext()
    private let edgeDetector = CIFilter(name: "CIEdges")
    private let colorControls = CIFilter(name: "CIColorControls")
    
    private init() {}
    
    struct FilterParams {
        var threshold1: CGFloat = 100.0  // Python: threshold1 = 100
        var threshold2: CGFloat = 200.0  // Python: threshold2 = 200
        var enableColorful: Bool = true
        var detectFaceOnly: Bool = true
        
        // HoughLinesP parameters to match Python code
        var houghThreshold: Int = 30      // Python: threshold = 30
        var minLineLength: Int = 80       // Python: minLineLength = 80
        var maxLineGap: Int = 5           // Python: maxLineGap = 5
    }
    
    /// Apply AI filter to transform face into colorful edge detection
    func applyAIFilter(to image: UIImage, params: FilterParams) -> UIImage? {
        guard let ciImage = CIImage(image: image) else { return nil }
        
        // Convert to grayscale for edge detection
        let grayscale = ciImage.applyingFilter("CIPhotoEffectNoir")
        
        // Apply Gaussian blur to reduce noise
        let blurred = grayscale.applyingGaussianBlur(sigma: 1.4)
        
        // Apply edge detection (Canny-like effect using CIEdges)
        edgeDetector?.setValue(blurred, forKey: kCIInputImageKey)
        edgeDetector?.setValue(params.threshold1 / 255.0, forKey: kCIInputIntensityKey)
        
        guard let edges = edgeDetector?.outputImage else { return nil }
        
        // Create black background
        let blackBackground = CIImage(color: CIColor.black)
            .cropped(to: ciImage.extent)
        
        var result: CIImage
        
        if params.enableColorful {
            // Apply colorful edge effect
            result = applyColorfulEdges(edges: edges, originalImage: ciImage)
        } else {
            // Simple white edges on black background
            result = edges
        }
        
        // Composite edges over black background
        let composite = result.composited(over: blackBackground)
        
        // Convert back to UIImage
        guard let cgImage = context.createCGImage(composite, from: composite.extent) else { return nil }
        return UIImage(cgImage: cgImage)
    }
    
    /// Apply colorful edges based on edge angles (similar to the Python/Java implementation)
    private func applyColorfulEdges(edges: CIImage, originalImage: CIImage) -> CIImage {
        // Detect lines using custom implementation
        let lines = detectLines(in: edges)
        
        // Create a new image for colored edges
        let extent = edges.extent
        UIGraphicsBeginImageContextWithOptions(extent.size, false, 1.0)
        guard let ctx = UIGraphicsGetCurrentContext() else {
            UIGraphicsEndImageContext()
            return edges
        }
        
        // Fill with black background
        ctx.setFillColor(UIColor.black.cgColor)
        ctx.fill(CGRect(origin: .zero, size: extent.size))
        
        if !lines.isEmpty {
            // Draw colored lines based on angle (matching Python algorithm exactly)
            for line in lines {
                // Python: a = (x1 - x2) / (y1 - y2)
                let x1 = line.start.x, y1 = line.start.y
                let x2 = line.end.x, y2 = line.end.y
                
                if y1 != y2 {
                    let a = (x1 - x2) / (y1 - y2)
                    // Python: b = round(math.degrees(math.atan(a))) + 90
                    let angleRadians = atan(a)
                    let degrees = round(angleRadians * 180.0 / .pi) + 90
                    
                    // Python: c = hsv_to_rgb(b, 255, 255)
                    let hsvColor = hsvToRgb(h: Int(degrees), s: 255, v: 255)
                    let color = UIColor(red: CGFloat(hsvColor.r)/255.0, 
                                      green: CGFloat(hsvColor.g)/255.0, 
                                      blue: CGFloat(hsvColor.b)/255.0, 
                                      alpha: 1.0)
                    
                    ctx.setStrokeColor(color.cgColor)
                    ctx.setLineWidth(3.0)
                    ctx.move(to: line.start)
                    ctx.addLine(to: line.end)
                    ctx.strokePath()
                }
            }
        } else {
            // If no lines detected, show rainbow gradient edges
            guard let cgImage = context.createCGImage(edges, from: edges.extent) else {
                UIGraphicsEndImageContext()
                return edges
            }
            
            let edgeImage = UIImage(cgImage: cgImage)
            
            // Apply rainbow gradient based on position
            for y in stride(from: 0, to: Int(extent.height), by: 2) {
                for x in stride(from: 0, to: Int(extent.width), by: 2) {
                    let point = CGPoint(x: x, y: y)
                    
                    // Check if edge exists at this point
                    if isEdgeAt(point: point, in: edgeImage) {
                        let hue = CGFloat((x + y) % 180) / 180.0
                        let color = UIColor(hue: hue, saturation: 1.0, brightness: 1.0, alpha: 1.0)
                        
                        ctx.setFillColor(color.cgColor)
                        ctx.fill(CGRect(x: CGFloat(x), y: CGFloat(y), width: 2, height: 2))
                    }
                }
            }
        }
        
        let coloredImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        guard let finalImage = coloredImage,
              let ciColoredImage = CIImage(image: finalImage) else {
            return edges
        }
        
        return ciColoredImage
    }
    
    /// Detect lines in edge image (simplified Hough transform)
    private func detectLines(in edgeImage: CIImage) -> [Line] {
        var lines: [Line] = []
        
        // Convert to CGImage for pixel access
        guard let cgImage = context.createCGImage(edgeImage, from: edgeImage.extent) else {
            return lines
        }
        
        let width = cgImage.width
        let height = cgImage.height
        
        // Simple line detection (simplified for performance)
        let minLineLength = 80
        let maxGap = 5
        
        // Scan for horizontal and vertical lines
        for y in stride(from: 0, to: height, by: 10) {
            var lineStart: CGPoint?
            var lineLength = 0
            
            for x in 0..<width {
                if isPixelBright(at: CGPoint(x: x, y: y), in: cgImage) {
                    if lineStart == nil {
                        lineStart = CGPoint(x: x, y: y)
                    }
                    lineLength += 1
                } else {
                    if let start = lineStart, lineLength > minLineLength {
                        lines.append(Line(start: start, end: CGPoint(x: x - 1, y: y)))
                    }
                    lineStart = nil
                    lineLength = 0
                }
            }
        }
        
        return lines
    }
    
    /// Check if pixel is bright (edge)
    private func isPixelBright(at point: CGPoint, in cgImage: CGImage) -> Bool {
        guard let dataProvider = cgImage.dataProvider,
              let data = dataProvider.data else {
            return false
        }
        
        let pixelData = CFDataGetBytePtr(data)
        let bytesPerRow = cgImage.bytesPerRow
        let bytesPerPixel = cgImage.bitsPerPixel / 8
        
        let pixelOffset = Int(point.y) * bytesPerRow + Int(point.x) * bytesPerPixel
        
        if pixelOffset < CFDataGetLength(data) {
            let brightness = pixelData?[pixelOffset] ?? 0
            return brightness > 128
        }
        
        return false
    }
    
    /// Check if edge exists at point
    private func isEdgeAt(point: CGPoint, in image: UIImage) -> Bool {
        guard let cgImage = image.cgImage else { return false }
        return isPixelBright(at: point, in: cgImage)
    }
    
    /// Line structure
    private struct Line {
        let start: CGPoint
        let end: CGPoint
    }
    
    /// HSV to RGB conversion matching Python cv2.cvtColor(HSV2BGR)
    private func hsvToRgb(h: Int, s: Int, v: Int) -> (r: Int, g: Int, b: Int) {
        let hNorm = CGFloat(h % 360) / 360.0
        let sNorm = CGFloat(s) / 255.0
        let vNorm = CGFloat(v) / 255.0
        
        let c = vNorm * sNorm
        let x = c * (1 - abs((hNorm * 6).truncatingRemainder(dividingBy: 2) - 1))
        let m = vNorm - c
        
        var r: CGFloat = 0, g: CGFloat = 0, b: CGFloat = 0
        
        let sector = Int(hNorm * 6)
        switch sector {
        case 0: (r, g, b) = (c, x, 0)
        case 1: (r, g, b) = (x, c, 0)
        case 2: (r, g, b) = (0, c, x)
        case 3: (r, g, b) = (0, x, c)
        case 4: (r, g, b) = (x, 0, c)
        default: (r, g, b) = (c, 0, x)
        }
        
        return (r: Int((r + m) * 255), g: Int((g + m) * 255), b: Int((b + m) * 255))
    }
}